contents/security/sqlInjection.md

```diff
@@ -0,0 +1,217 @@
+or other important information used by the website. Databases are usually made of many "tables" 
+which are organised in rows and columns. Each row is a separate entry in the table, and 
+each column is a specific parameter which can be used by the entry.
+
+As an example, a table in a database may look like this:
+
+
+<table>
+	<tr>
+		<th>UserId</th>
+		<th>Username</th>
+		<th>Password</th>
+	</tr>
+	<tr>
+		<td>1</td>
+		<td>Admin</td>
+		<td>123456</td>
+	</tr>
+	<tr>
+		<td>2</td>
+		<td>Alice</td>
+		<td>pw1234</td>
+	</tr>
+	<tr>
+		<td>3</td>
+		<td>...</td>
+		<td>...</td>
+	</tr>
+</table>
+
+
+SQL is used to interact with the database by sending "queries" which the database responds to. 
+Some common SQL commands used include: 
+* SELECT - retrieves information from a table
+* UPDATE - changes information from a table
+* INSERT INTO - adds a new entry into a table
+* DELETE FROM - deletes information from a table
+
+Websites combine SQL syntax with user input to query the underlying database of the website. 
+For example, blog websites may ask for a username and password from users for them to log 
+in to the site. This input is used to create the SQL statement that will be sent to the database. 
+A website which is vulnerable to SQL injection attacks may construct the SQL query like this:
+
+```sql
+String query = "SELECT * FROM Users WHERE Username = ‘" + input_username +"’ 
+                AND Password = ‘" + input_password + "’";
+```
+
+As an example, if a user entered their username, "Alice", and password, "pw1234" into the website 
+to try to gain access:
+
+```
+Username: Alice
+Password: pw1234
+```
+
+The constructured query would look like this:
+
+```sql
+SELECT * FROM Users WHERE Username = ‘Alice’ 
+AND Password = ‘pw1234’
+```
+
+This query would find all entries in the "Users" table in the database, and return any 
+entries where the "Username" is "Alice" and the "Password" is "pw1234". If the previous 
+table was searched, it would return the details of the second row as the result. Since there was 
+a result returned, the website would be able to tell that a legitimate username and password 
+combination was entered since the query requires that both are matched to retrieve the data.
+Thus, the website would know that the user is legitimate, and the user would be allowed to log 
+into the site. On the other hand, if no result was returned by the database, the website would know 
+that the username and password combination does not match any of the users in the database, and would 
+deny access to the person trying to log in.
+
+
+However, are users limited to just filling in legitimate username and password strings? 
+Some websites directly use the information given by the user without any validation or 
+sanitization, thus the users are able to write code as input for their own malicious 
+purposes. For example, users are able to add parameters to the query.
```

Comment:
Sanitation is a technical term used here as if the reader is supposed to know what it means.